app/models/category.rb

```diff
@@ -0,0 +1,21 @@
+class Category < ApplicationRecord
+  has_many :sub_categories, class_name: "Category", foreign_key: "parent_id", dependent: :destroy
+  belongs_to :parent, class_name: "Category", optional: true
+
+  has_many :products, dependent: :restrict_with_error
+  has_many :sub_category_products,  through: :sub_categories,  source: :products
+
+  validates :name, presence: true
+  
+  validates_with UniqueRootCategoryName, if: :isNamePresent?
```

Comment:
no need of custom validation here. use uniqueness with conditions

Reply:
validates :name, uniqueness: { scope: :parent_id, message: 'name of root category should be unique' }, allow_blank: true, unless: :isParentPresent?